File: Android/shared/src/main/kotlin/shared/dto/objects/Timetable.kt
@file:UseSerializers(UUIDSerializer::class, DateSerializer::class, TimetableOwnerSerializer::class)

package shared.dto.objects

import com.maestri.sdk.sources.shared.serializers.DateSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import shared.dto.enums.AppointmentType
import shared.dto.enums.TimetableOwner
import shared.dto.primitives.Schedule
import shared.dto.protocols.Parametable
import shared.dto.protocols.Responsable
import shared.serializers.TimetableOwnerSerializer
import shared.serializers.UUIDSerializer
import java.util.*

typealias Intervals = List<DateInterval>

object Timetable {
    data object Parameters {
        data object Create {
            @Serializable
            data class Pattern(
                val schedule: Schedule.Pattern,
                val startAt: Date,
                val endAt: Date?,
            ) : Parametable()

            @Serializable
            data class Flexible(
                val workDays: Map<Date, Schedule.Day>
            ) : Parametable()
        }

        @Serializable
        data class SearchSlot(
            val appointmentType: AppointmentType,
            val customerId: UUID?,
        ) : Parametable()

        @Serializable
        data class Retrieve(
            val owners: List<TimetableOwner>,
            //Идеально отправлять в salon time zone с 00:00-00:00 чтобы были только дни
            val period: DateInterval,
        ) : Parametable()
    }

    data object Responses {
        // Используется для возвращения найденных слотов на которые можно записаться
        @Serializable
        data class Slots(
            val intervals: Intervals,
            val timeZoneId: String,
        ) : Responsable

        @Serializable
        data class Schedule(
            val owner: TimetableOwner,
            //для недели 7 дней для месяца 28-31
            val intervals: Intervals,
            val timeZoneId: String,
        ) : Responsable
    }
}

File: Node/src/dto/objects/Timetable.ts
import { DateInterval } from "../tsPrimitives/DateInterval"
import { UUID } from "../tsPrimitives/UUID"
import { Schedule } from "../primitives/Schedule"
import { TimetableOwner } from "../enums/TimetableOwner"
import { AppointmentType } from "../enums/AppointmentType"
import { IsTimeZone, ValidateNested } from "class-validator"
import { IsoDate } from "../tsPrimitives/IsoDate"

export namespace Timetable {

    export namespace Parameters {

        export namespace Create {

            export class Pattern {
                startAt: IsoDate
                endAt?: IsoDate

                @ValidateNested()
                schedule: Schedule.Pattern

                constructor(
                    schedule: Schedule.Pattern, 
                    startAt: IsoDate, 
                    endAt?: IsoDate
                ) {
                    this.schedule = schedule
                    this.startAt = startAt
                    this.endAt = endAt
                }
            }

            export class Flexible {
                @ValidateNested()
                workDays: Map<Date, Schedule.Day>

                constructor(
                    workDays: Map<Date, Schedule.Day>
                ) {
                    this.workDays = workDays
                }
            }
        }

        export class SearchSlot {
            appointmentType: AppointmentType

            @ValidateNested()
            customerId?: UUID

            constructor(
                appointmentType: AppointmentType,
                customerId?: UUID
            ) {
                this.appointmentType = appointmentType
                this.customerId = customerId
            }
        }

        export class Retrieve {
            @ValidateNested()
            owners: TimetableOwner[]

            @ValidateNested()
            period: DateInterval

            constructor(
                owners: TimetableOwner[],
                period: DateInterval
            ) {
                this.owners = owners
                this.period = period
            }
        }
    }

    export namespace Responses {

        type Intervals = DateInterval[]

        export class Slots {
            @ValidateNested()
            intervals: Intervals

            @IsTimeZone()
            timeZoneId: string

            constructor(
                intervals: Intervals,
                timeZoneId: string
            ) {
                this.intervals = intervals
                this.timeZoneId = timeZoneId
            }
        }

        export class Schedule {
            @ValidateNested()
            owner: TimetableOwner

            @ValidateNested()
            intervals: Intervals
            
            @IsTimeZone()
            timeZoneId: string

            constructor(
                owner: TimetableOwner,
                intervals: Intervals,
                timeZoneId: string
            ) {
                this.owner = owner
                this.intervals = intervals
                this.timeZoneId = timeZoneId
            }
        }
    }
}